[cmake][addons] automate addon.xml version set

AlwinEsch opened the pull request:
## Description
This automate the set of versions on addon.xml in kodi and also allow
set of them on addon itself.

On a Version change need only the number inside the Header "version.h"
replaced, everything else becomes then set from cmake.

<!--- Provide a general summary of your change in the Title above -->

**WARNING:** A addon without changed addon.xml does not work anymore!

After them works the new version.h way for addon.xml's

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

MartijnKaijser commented:
You need to keep the ABI set because it would now simply accept any Addon from before v18

AlwinEsch commented:
Understand but we must come to a point where it becomes cleaned up. If it becomes more and more is it really hard to know what is where.

I look to add a check and if a for kodi unknown Id e.g. "xbmc.pvr" is found becomes it inhibited.

AlwinEsch commented:
Have seen was already a check there if during load a unknown depend is found on xml who brings this:
![bildschirmfoto vom 2017-04-29 13-20-31](https://cloud.githubusercontent.com/assets/6879739/25555154/1b35b622-2cdf-11e7-99e4-c06c28bfea78.png)

With the other commit becomes this added:
![bildschirmfoto vom 2017-04-29 13-18-43](https://cloud.githubusercontent.com/assets/6879739/25555156/2b275fe0-2cdf-11e7-8c05-3757c5853df6.png)

AlwinEsch commented:
@MartijnKaijser have added a addon-blacklist.xml as opposite of addon-manifest.xml. If something is on addon.xml (his ID or a dependency) and inside blacklist becomes the addon not loaded.

Is this a acceptable way? If a dependency ID becomes changed can be the old added there.

stefansaraev commented:
hum, not sure why would need such a blacklist. if a specific dependency addon id does not exist anymore, then kodi should refuse to load the addon anyway, right ? at least it is doing so for python addons

AlwinEsch commented:
Thought about the case that someone install a new version without uninstall of the old where the no more supported parts still present.

MartijnKaijser commented:
I do not like the blacklist idea. That's what the whole ABI is about and you removed it.

stefansaraev commented:
I'd call that "broken installation". I can think of only few scenarios where this can happen:
1. user is not using his distro package manager but doing make install on top of his old installation. many other things may break in this case, anyway. user is supposed to know what he is doing when building from source.
2. windows. in case installer does not care to clean up old system dir. if so - that's a bug in the windows installer

AlwinEsch commented:
Then please give me a alternative, since 2 years I want to cleanup the whole binary add-on system to bring support for new parts and allow a binary like a Python for everything.

Even I look at the mixed one in addon-manifest.xml not quite through and can not directly assign parts.

MartijnKaijser commented:
You also need to update https://github.com/xbmc/xbmc/blob/71837f8bc620b264205e627704d82ad059a89921/xbmc/addons/Addon.cpp because you're breaking repo.

AlwinEsch commented:
The parts there have nothing to do with the dependencies, as example there is for PVR addon itself `xbmc.pvrclient` and on manifest it was `xbmc.pvr` for dependency.

AlwinEsch commented:
A dependency addon is a fake addon with unknown type who is normally not visible inside Kodi and have no dependencies to Kodi itself.

AlwinEsch commented:
Have removed the commit with the blacklist

AlwinEsch commented:
As info this request is only be related to depends of binary addon's. For everything else e.g. Python or Skin is nothing changed.

stefansaraev commented:
works for me.

AlwinEsch commented:
A question to the linked addon requests here with "change to automatic interface version set". Is a automatic set of version during compile OK or not?

This have for me a good and a bad part.
The good: Addon need on changed only a recompile (if no bigger changed)
The bad: Is a bit hard to understand in commit history.

If not good for you then I remove the auto set there on addon request's.

stefansaraev commented:
> If not good for you then I remove the auto set there on addon request's.

I have no opinion on this, so whatever you or the team think is better.

tamland commented:
Can you please hold a moment, and explain your end goal here and what trying to achieve with these version changes PRs other than 'the new version.h way'? Because I have no idea what it is from just looking at these incremental refactoring..

For example,
Why do "types" suddenly have versions, and what does that even mean? We never had this before.
Why are there now two compatibility checks, one done by the geneal addon system and a custom one to binary addon?
For other all other addons, runtime api compatibility have previously been solved by using `<require>`, so lastly: why do we need yet-another-custom-binary-addon-solution in addition to that?

tamland commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -776,16 +777,27 @@ bool CAddonMgr::DisableAddon(const std::string& id)
+  if (!IsCompatible(*addon))
+  {
+    CLog::Log(LOGERROR, "Add-on '%s' is incompatible against Kodi", addon->ID().c_str());
+    CEventLog::GetInstance().AddWithNotification(EventPtr(new CNotificationEvent(addon->Name(), 24152, EventLevel::Error)));
```
As I said before, please don't put gui logic all over the core systems. This method is used for other things, like enabling dependencies. Showing dialogs, notification popups and things like that should be handled in CGUIDialogAddonInfo.

AlwinEsch replied:
There must be a user informed that after his Kodi update the addon is no more working. This is the best place where it can be done and with this `CEventLog::GetInstance().AddWithNotification` is the running system not touched.

ksooo commented:
> Can you please hold a moment, and explain your end goal here

 Yes, please. I recently see lots of changes and asking myself what problems you're trying to solve. What gets better? Why do we need/want this?

AlwinEsch commented:
There was always a version check on binary addons by addon.xml and with code this was before criss and cross defined by his headers and in addon.xml of depends in Kodi, every addon interface system have own calls to check them.

Problem of the old is that sometimes the version on code becomes changed but not the addon.xml or the addon.xml and not the code, some version defines in headers was never used.
The version functions in code are currently direct on his related type function tables, if now PVR or everything else becomes bigger changes can his function table no more work. Thats why I want a global fixed function.
Further bring the version.h the possibility to set the addon.xml by his values and on changes only this must be updated.
Also becomes this version check done by his global addon classes and the same code copies from all binary addon types is no more needed.

On next request becomes e.g. the:
```cpp
    const char*  (__cdecl* GetPVRAPIVersion)(void);
    const char*  (__cdecl* GetMininumPVRAPIVersion)(void);
    const char*  (__cdecl* GetGUIAPIVersion)(void);
    const char*  (__cdecl* GetMininumGUIAPIVersion)(void);
```
removed (also on other addon types).

AlwinEsch commented:
To read the versions further from code (and not only from addon.xml) was also requested from @FernetMenta to prevent code problems all the time.

ksooo commented:
> To read the versions further from code (and not only from addon.xml) ...

 Isn't reading the version from one location (either addon.xml OR code) enough?

FernetMenta commented:
crappy cpluff is not reliable. that's why @opdenkamp indroduced code checks

AlwinEsch commented:
Also a reason to have still on addon.xml is to prevent load of any addon dll only to check his version.

On end if binary addon is really used becomes then with Dll function confirmed that code and xml are synchronized.

ksooo commented:
> crappy cpluff is not reliable. that's why @opdenkamp indroduced code checks

 I remember Lars saying this was because addon devs are sometimes to dumb to maintain addon.xml properly (they tend to forget) and then Kodi crashes if it tries to load the incompatible addon...

 So, the goal is to get rid of the code checks, imo.

FernetMenta commented:
goal is it get rid of cpluff

hudokkow commented:
Alwin seems to be having some difficulty explaining what he means/is aiming in English. Perhaps it's easier for you guys to communicate in German? Just for this one PR? Or internally?

Please don't be offended, Alwin. Language barrier sometimes gets in the way of correctly explaining what one is trying to achieve.

tamland commented:
Agree with ksooo, but we should *also* get rid of the whole 'fake addons' thing. It's a cute idea, but doesn't really work. They're already treated specially multiple places so atm they are pointless to keep as fake addons. Imo we should hard code api versions in source (but keep the same semantics of specifying dependencies and not loading incompatible ones).

Then, if you want, you could do this post load version check without having to update multiple places (but honestly I'm not sure if there's any point..)

ksooo commented on addons/kodi.binary.global.main/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.global.main" version="@ADDON_GLOBAL_VERSION_MAIN@" provider-name="Team KODI">
```
"Team KODI" => "Team Kodi" (for consistency)

ksooo commented on addons/kodi.binary.instance.adsp/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.instance.adsp" version="@ADDON_INSTANCE_VERSION_ADSP@" provider-name="Team KODI">
```
"Team KODI" => "Team Kodi" (for consistency)

ksooo commented on addons/kodi.binary.instance.game/addon.xml.in:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<addon id="kodi.game" version="1.0.29" provider-name="Team-Kodi">
-	<backwards-compatibility abi="1.0.29"/>
+<addon id="kodi.binary.instance.game" version="@ADDON_INSTANCE_VERSION_GAME@" provider-name="Team-Kodi">
```
"Team-Kodi" => "Team Kodi" (without the dash, for consistency)

ksooo commented on addons/kodi.binary.instance.peripheral/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.instance.peripheral" version="@ADDON_INSTANCE_VERSION_PERIPHERAL@" provider-name="Team-Kodi">
```
"Team-Kodi" => "Team Kodi" (without the dash, for consistency)

ksooo commented on addons/kodi.binary.instance.pvr/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.instance.pvr" version="@ADDON_INSTANCE_VERSION_PVR@" provider-name="Team-Kodi">
```
"Team-Kodi" => "Team Kodi" (without the dash, for consistency)

ksooo commented on addons/kodi.binary.instance.screensaver/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.instance.screensaver" version="@ADDON_INSTANCE_VERSION_SCREENSAVER@" provider-name="Team-Kodi">
```
"Team-Kodi" => "Team Kodi" (without the dash, for consistency)

ksooo commented on addons/kodi.binary.instance.vfs/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.instance.vfs" version="@ADDON_INSTANCE_VERSION_VFS@" provider-name="Team-Kodi">
```
"Team-Kodi" => "Team Kodi" (without the dash, for consistency)

ksooo commented on addons/kodi.binary.instance.visualization/addon.xml.in:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<addon id="kodi.binary.instance.visualization" version="@ADDON_INSTANCE_VERSION_VISUALIZATION@" provider-name="Team-Kodi">
```
"Team-Kodi" => "Team Kodi" (without the dash, for consistency)

ksooo commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -68,6 +68,19 @@ macro (build_addon target prefix libs)
   # if there's an addon.xml.in we need to generate the addon.xml
   if(EXISTS ${PROJECT_SOURCE_DIR}/${target}/addon.xml.in)
+    # Set defines used on addon.xml.in and readed from versions.h
+    # To set versions about add-on parts automatic
```
"Set defines used in addon.xml.in and read from versions.h to set add-on version parts automatically" ?

ksooo commented on cmake/scripts/common/GenerateVersionedFiles.cmake:
```diff
@@ -13,6 +11,30 @@ function(generate_versioned_file _SRC _DEST)
+# add-on xml's
+file(GLOB ADDON_XML_IN_FILE ${CORE_SOURCE_DIR}/addons/*/addon.xml.in)
+foreach(loop_var ${ADDON_XML_IN_FILE})
+  # prevent 'xbmc.json', becomes done from 'xbmc/interfaces/json-rpc/schema/CMakeLists.txt'.
```
", becomes done" => "; will be obtained from"

ksooo commented on cmake/scripts/common/Macros.cmake:
```diff
@@ -629,22 +627,29 @@ macro(core_find_versions)
-  string(REGEX REPLACE ".*\"(.*)\"" "\\1" guilib_version_min ${guilib_version_min})
+
+  # Set defines used on addon.xml.in and readed from versions.h
+  # To set versions about add-on parts automatic
```
"Set defines used in addon.xml.in and read from versions.h to set add-on version parts automatically" ?

ksooo commented on cmake/scripts/common/Macros.cmake:
```diff
@@ -629,22 +627,29 @@ macro(core_find_versions)
+
+  # Set defines used on addon.xml.in and readed from versions.h
+  # To set versions about add-on parts automatic
+  # This part is near equal with them on "AddonHelpers.cmake", except the place of versions.h
```
"This part is nearly identical to "AddonHelpers.cmake", except location of versions.h" ?

ksooo commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -400,31 +390,5 @@ ADDON_STATUS CAddonDll::TransferSettings()
-{
-  /* check the API version */
-  const char* kodiVersion = kodi::addon::GetTypeVersion(type);
-  const char* addonVersion = m_pDll->GetAddonTypeVersion(type);
```
Shouldn't we remove GetAddonTypeVersion from Addon API, now that we do no longer check that data?

AlwinEsch replied:
Thought lo leave it in, in case it will be needed in the future and the old addons allow the check.

Because it is made in the headers and the addon is not directly affected.

Thanks for your other comments, I replace the strings there.

ksooo commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -776,16 +777,27 @@ bool CAddonMgr::DisableAddon(const std::string& id)
+  
+  if (!IsCompatible(*addon))
+  {
+    CLog::Log(LOGERROR, "Add-on '%s' is incompatible against Kodi", addon->ID().c_str());
```
""Add-on '%s' is not compatible with Kodi"

AchimTuran commented:
> Thought lo leave it in, in case it will be needed in the future and the old addons allow the check.

As far as I understand you completely, you wanna remove add-on version code checks in code and do all checks in the xml file. If it is a real clean-up you should remove those checks in code completely.

ksooo commented:
> As far as I understand you completely, you wanna remove add-on version code checks in code and do all checks in the xml file. If it is a real clean-up you should remove those checks in code completely.

+1

AlwinEsch commented:
Request is updated, now becomes every depend of binary addon automatically added (addon.xml.in)

Here on addon.xml.in:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<addon
  id="pvr.vdr.vnsi"
  version="3.0.3"
  name="VDR VNSI Client"
  provider-name="FernetMenta, Team Kodi">
  <requires>@ADDON_DEPENDS@</requires>
  <extension
    point="xbmc.pvrclient"
    library_@PLATFORM@="@LIBRARY_FILENAME@"/>
```
and here the returned on addon.xml:
```cpp
<?xml version="1.0" encoding="UTF-8"?>
<addon
  id="pvr.vdr.vnsi"
  version="3.0.3"
  name="VDR VNSI Client"
  provider-name="FernetMenta, Team Kodi">
  <requires>
<import addon="kodi.binary.global.main" version="1.0.0"/>
<import addon="kodi.binary.global.gui" version="5.11.0"/>
<import addon="kodi.binary.instance.pvr" version="5.2.2"/></requires>
  <extension
    point="xbmc.pvrclient"
    library_linux="pvr.vdr.vnsi.so.3.0.3"/>
```

This goes with a check over all on addons CMakeLists.txt defined source parts and check for used included (In case the headers are not defined in txt becomes everything with *.h* checked on addon.
It need around 2 seconds, but after worth are all used parts known.
To know the used parts are they defined in "version.h"

AchimTuran commented:
@AlwinEsch 
> To know the used parts are they defined in "version.h"

Maybe add a link to your latest version/vision of the file "version.h", so others can see where the journy will end.

AlwinEsch commented:
For the cmake needed Versions to read are defined here: https://github.com/AlwinEsch/kodi/blob/87af7eeda88908baa931d49a4a3d3ee370cbff32/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L42-L128

Maybe on end if everything is done from me for the complete rework, comes a addon-dependencies.xml who then contains the versions and this becomes then added to a `versions.h.in`, so that only this need to change.

wsnipex commented on CMakeLists.txt:
```diff
@@ -76,6 +76,34 @@ endif()
+  endforeach()
+  unset(xml_name)
+endforeach()
+
```
please move this whole block further down. Or even better, move it to a function or macro.

wsnipex commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -35,7 +35,81 @@ endmacro()
+        list(APPEND USED_SOURCES ${PROJECT_SOURCE_DIR}/${hdr_file})
+      endforeach()
+    else()
+      message(AUTHOR_WARNING "You have not defined used headers in cmake, scanning all addon included files!")
```
"Header files not defined in your CMakeLists.txt. Please consider defining ${prefix}_HEADERS as list of all headers used by this addon. Falling back to recursive scan for *.h"

AlwinEsch commented:
@wsnipex have changed it related to your commits, thanks a lot!

wsnipex commented on cmake/scripts/common/GenerateVersionedFiles.cmake:
```diff
@@ -13,6 +11,30 @@ function(generate_versioned_file _SRC _DEST)
+
+  file(READ "${source_dir}/addon.xml.in" file_content)
+  string(CONFIGURE "${file_content}" file_content @ONLY)
+  file(WRITE "${dest_dir}/addon.xml" "${file_content}")
```
configure_file()?

wsnipex commented on cmake/scripts/common/Macros.cmake:
```diff
@@ -600,9 +600,7 @@ endfunction()
-# The following variables are set from versions.h:
-#   guilib_version - current ADDONGUI API version
-#   guilib_version_min - minimal ADDONGUI API version
+# About for addon's set variables see defines in "versions.h"
```
Set various variables defined in "versions.h"

AlwinEsch commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -35,7 +35,81 @@ endmacro()
+    file(STRINGS ${KODI_INCLUDE_DIR}/versions.h BIN_ADDON_PARTS)
+    foreach(loop_var ${BIN_ADDON_PARTS})
+      # Only pass strings with "#define" from versions.h
+      if(loop_var MATCHES "#define")
```
The regex is still Greek to me! :-D 
Has tried several strings and not found the right to become only the parts with "#define".

AlwinEsch commented:
jenkins build addons please

AlwinEsch commented:
jenkins build this with addons please

Rechi commented:
@AlwinEsch this needs to rebased

AlwinEsch commented:
I know, was only to check the addons are supported. Still comes cleanup from them and want to make sure it works already for jenkins.

Is a bit crazy that all on all addons 5-6 different ways how headers or sources are added.

AlwinEsch commented:
Has found some parts and next try, sorry jenkins.

jenkins build this with addons please

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
Has looked now, if it is done with `externalproject_add` does it work, but here without external not (https://github.com/notspiff/audiodecoder.modplug/blob/master/depends/windows/modplug/CMakeLists.txt)

AlwinEsch commented:
jenkins build this with addons please

notspiff commented:
yes, it's the externalproject which handles the patchery. this CMakeLists.txt is just used to do the actual build of the already-patched sources.

AlwinEsch committed:
[cmake][addons] automate addon.xml version set

This automate the set of versions on addon.xml in kodi and also allow
set of them on addon itself.

On a Version change need only the number inside the Header "version.h"
replaced, everything else becomes then set from cmake.

AlwinEsch commented:
If OK then I merge this request on next night and create requests for all binary addons to match them. Jenkins is mostly OK now.

AlwinEsch commented:
jenkins build this with addons please (the last time)

hudokkow commented:
jenkins build this with addons please

tamland commented:
Still gui code in internal AddonMgr methods which I asked to be move into the addon dialog.

AlwinEsch commented:
Yes @tamland, problem is that related error can be come automatic after a Kodi Base update during his start. The addon dialog is still closed then. If now 20 binary addons are outdated from his versions becomes it problematic to handle them with dialogs.

There is a point who need improved in future, but so long binaries are not so available from a repo like the python addons, is it not possible to do a right addon update during Kodi start.

notspiff commented:
hope you didn't hope to ref in any of the pr's - all ref'd here merged from my side.
